subsys/net/ip/tcp.c

```diff
@@ -2042,16 +2050,24 @@ NET_CONN_CB(tcp_established)
 
 	net_context_set_appdata_values(pkt, IPPROTO_TCP);
 
-	data_len = net_pkt_appdatalen(pkt);
-	if (data_len > net_tcp_get_recv_wnd(context->tcp)) {
-		NET_ERR("Context %p: overflow of recv window (%d vs %d), "
-			"pkt dropped",
-			context, net_tcp_get_recv_wnd(context->tcp), data_len);
-		return NET_DROP;
+	if (!context->tcp->zero_wnd_sent) {
+		data_len = net_pkt_appdatalen(pkt);
+		if (data_len > net_tcp_get_recv_wnd(context->tcp)) {
+			NET_ERR("Context %p: overflow of recv window "
+				"(%d vs %d), pkt dropped",
+				context, net_tcp_get_recv_wnd(context->tcp),
+				data_len);
+			return NET_DROP;
+		}
+	} else {
+		/* Zero window probe expected. */
+		data_len = 0;
+		force_ack = true;
```

Comment:
Yes, but this does more changes than necessary. In this function, we already have a piece of code which force-resends an ACK, here: https://github.com/zephyrproject-rtos/zephyr/pull/7867/files#diff-a278935f8d5d5e0b4424ca3e1182d48cR1965 . Just goto'ing there is enough.

Reply:
From what I've seen it will work only if we receive SEQ lower then recently sent ACK. In case of ZWP they're equal (just a next byte of data).

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/ip/tcp.c#L1957 calls send_acl() with force_ack set to true, just like you try to do with adding extra variable.